packages/xo-server/src/xapi-stats.js

```diff
@@ -17,383 +35,165 @@ const RRD_STEP_FROM_STRING = {
   days: RRD_STEP_DAYS,
 }
 
+// points = intervalInSeconds / step
 const RRD_POINTS_PER_STEP = {
   [RRD_STEP_SECONDS]: 120,
   [RRD_STEP_MINUTES]: 120,
   [RRD_STEP_HOURS]: 168,
   [RRD_STEP_DAYS]: 366,
 }
 
-export class XapiStatsError extends BaseError {}
-
-export class UnknownLegendFormat extends XapiStatsError {
-  constructor (line) {
-    super('Unknown legend line: ' + line)
-  }
-}
-
-export class FaultyGranularity extends XapiStatsError {}
-
 // -------------------------------------------------------------------
 // Utils
 // -------------------------------------------------------------------
 
-// Return current local timestamp in seconds
-function getCurrentTimestamp () {
-  return Date.now() / 1000
+async function getServerTimestamp (xapi, hostRef) {
+  const serverLocalTime = await xapi.call('host.get_servertime', hostRef)
+  return Math.floor(parseDateTime(serverLocalTime).getTime() / 1e3)
 }
 
 function convertNanToNull (value) {
   return isNaN(value) ? null : value
 }
 
-async function getServerTimestamp (xapi, host) {
-  const serverLocalTime = await xapi.call('host.get_servertime', host.$ref)
-  return Math.floor(parseDateTime(serverLocalTime).getTime() / 1000)
+// Return current local timestamp in seconds
+function getCurrentTimestamp () {
+  return Date.now() / 1000
 }
 
 // -------------------------------------------------------------------
 // Stats
 // -------------------------------------------------------------------
 
-function getNewHostStats () {
-  return {
-    cpus: [],
-    pifs: {
-      rx: [],
-      tx: [],
-    },
-    load: [],
-    memory: [],
-    memoryFree: [],
-    memoryUsed: [],
+const computeValues = (dataRow, legendIndex, transformValue = identity) =>
+  map(dataRow, ({ values }) =>
+    transformValue(convertNanToNull(values[legendIndex]))
+  )
+
+// It browse the object in depth and initialise it's properties
+// The targerPath can be a string or an array containing the depth
+// targetPath: [a, b, c] => a.b.c
+const getValuesFromDepth = (obj, targetPath) => {
+  if (isString(targetPath)) {
+    return (obj[targetPath] = [])
   }
+
+  forEach(targetPath, (path, key) => {
+    if (obj[path] === undefined) {
+      obj = obj[path] = targetPath.length - 1 === key ? [] : {}
+      return
+    }
+    obj = obj[path]
+  })
+  return obj
```

Comment:
Why not using `lodash.get`?

Reply:
because i need also initialise values in depth if they are undefined.

Reply:
I see.